### ISSUE TITLE
Cache performance tests

### ISSUE THREAD

**Issue by npcottaris** [isetbio/RemoteDataToolbox#34]
Hello,

I did a little  testing by comparing the loading times for a dataset of hyperspectral images totaling 6.02 GB. Here are my findings:

```
* Loading from remote source: 1st time     |   144 seconds
* Loading from remote source: 2nd time     |   134 seconds
* Loading from remote source: 3rd time     |   135 seconds
* Loading from local disk: 1st time        |    55 seconds
* Loading from local disk: 2nd time        |    54 seconds
* Loading from local disk: 3rd time        |    55 seconds
```

The remote loading was done via

```
client = RdtClient('isetbio');
client.crp(remotePath);
client.readArtifac(filename, 'type', filetype)
```
- The performance improvement during remote loading following the initial access is not that dramatic and does not approach the local loading performance.
- Do we have to configure any settings to enable caching in the RDT, or is this done automatically?
- If there is a cache directory, where is it located?

Nicolas

**Comment by wandell** [comment]
In reading through the code, I didn't see any local caching.  Do you think it has been implemented somewhere?

I do believe that the shift to the linear model representation will speed up the downloads a lot.

**Comment by DavidBrainard** [comment]
My understanding is that gradle handles local caching, with a cache directory that sits somewhere by default and which you can move around with some setting in a JSON file.  But it is possible that it isn't turned on by default.

I will review all of the recent RDTB traffic with Ben when I Skype with him tomorrow.

**Comment by npcottaris** [comment]
Yes, there is no cache specification in the code. Echoing David's response, I also thought that caching was transparent to the user.

**Comment by benjamin-heasly** [comment]
Woah, thanks for the numbers.  I am surprised that the there is not a big cache effect.

I'll try to reproduce and investigate on my end.

Gradle is supposed to do the caching for us and it should be always on.  So something must be confusing Gradle.

**Comment by benjamin-heasly** [comment]
The default cache folder is usually something ugly like `/home/ben/.gradle/caches/modules-2/files-2.1/`.  When you read an artifact, you can see this in the returned `localPath`.  For example:

```
>> [data, artifact] = client.readArtifact( ... );
>> disp(artifact)

      artifactId: 'demo-image'
     description: 'This is a test image with random-colored vertical stripes.'
       localPath: '/home/ben/.gradle/caches/modules-2/files-2.1/project-demo/demo-image/42/ae1df0594d8d6f192822f…'
            name: 'Dr. Stripes'
      remotePath: 'project-demo'
    repositoryId: ''
            type: 'png'
             url: 'http:/52.32.77.154/repository/demo-repository/project-demo/demo-image'
         version: '+'
```

You can specify a different cache folder with a configuration parameter called `cacheFolder`.  In JSON it would look like this:

```
{
    "serverUrl": "http://52.32.77.154/",
    "repositoryUrl": "http://52.32.77.154/repository/demo-repository/",
    "username": "guest",
    "password": "",
    "repositoryName": "demo-repository",
    "cacheFolder": "/home/ben/Desktop/alternate-cache-folder"
}
```

And then you should get:

```
       localPath: '/home/ben/Desktop/alternate-cache-folder/caches/modules-2/files-2.1/project-demo/demo-image/42/ae1df0594d8d6f192822f…'
```

**Comment by npcottaris** [comment]
Hi Ben,

My localPath is `/Users/nicolas/.gradle/caches/modules-2/files-2.1/ ...`
Checked in there and indeed the files are there.

I also did another test with a smaller dataset (1.32 GB). Same results:

```
* Loading from remote source: 1st time     |   24.81seconds
* Loading from remote source: 2nd time     |   20.32 seconds
* Loading from remote source: 3rd time     |   20.23 seconds
* Loading from local disk: 1st time        |    7.50 seconds
* Loading from local disk: 2nd time        |    7.38 seconds
* Loading from local disk: 3rd time        |    7.36 seconds
```

**Comment by benjamin-heasly** [comment]
Hi Nicolas,

OK, interesting.

What I'm seeing on my end is:
- Fetching with gradle is slower that just reading a local mat-file, about 2 seconds rather than 50ms.  Not ideal.  I expect some overhead from gradle, maybe I can squeeze it down somehow.
- But I still see a big cache effect, about 23 seconds the first time vs 2 seconds after that.

So we are doing something different.

I'll try to make my test script portable and send it to you.  Then would you want to run it on your end and see what happens?

**Comment by benjamin-heasly** [comment]
Here is my test script.  Since you have RemoteDataToolbox and isetbio, I _think_ you can just run it.

[cacheTests.m.zip](https://github.com/isetbio/RemoteDataToolbox/files/69870/cacheTests.m.zip)

Sorry about the zip.  GihHub would not take a `.m` file.

Here is my output:

```
fetch times
   35.9274         0         0

   35.9274    1.5574         0

   35.9274    1.5574    1.5709

local load times
    0.0703         0         0

    0.0703    0.0578         0

    0.0703    0.0578    0.0573
```

**Comment by benjamin-heasly** [comment]
I can report that one of our existing Gradle optimizations seems to be working.  This is the [Gradle Daemon](https://docs.gradle.org/current/userguide/gradle_daemon.html).  Basically, this lets us suffer Gradle startup times once, then reuse the same Gradle process afterwards.

The output above used the daemon.  I turned off the daemon and the same fetch times each became about 2 seconds slower:

```
40.5942    3.5770    3.6729
```

Well, this doesn't get us further than where we started, but it could be worse!

For those interested, we invoke the Daemon [here](https://github.com/isetbio/RemoteDataToolbox/blob/master/external/gradle/gradleFetchArtifact.m#L86) and [here](https://github.com/isetbio/RemoteDataToolbox/blob/master/external/gradle/gradlePublishArtifact.m#L92).

**Comment by npcottaris** [comment]
Hi Ben, I can confirm that running you test produces similar results on my machine.

```
fetch times
    5.8078         0         0
    5.8078    1.1265         0
    5.8078    1.1265    1.1013

local load times
    0.0369         0         0
    0.0369    0.0365         0
    0.0369    0.0365    0.0367
```

Modifying your script to download hyper spectral images I get the following:

```
Fetching from archiva for repeat 1 : 687.621861 seconds.
Fetching from archiva for repeat 2 :  13.330578 seconds.
Fetching from archiva for repeat 3 :  13.452849 seconds.
Fetching from local disk:              6.003912 seconds.
```

So there is a big effect of caching. I now believe that the results I reported earlier were for fetching data from Gradle's cache and fetching them from the disk. I must have downloaded the files one of the previous days, before I started examining cache performance, and I had not wiped out the cache directory, thinking that the data were not stored locally. Sorry about this !

**Comment by benjamin-heasly** [comment]
No problem!  I'm glad you brought this up anyway because I it caused us to put some numbers on the cache and the basic Gradle overhead.

I think Gradle overhead is something we should keep in mind going forward.

**Comment by DavidBrainard** [comment]
It might be worth someone at Stanford running the timing test script.  That will give us some sense of how much we might want location services.

It also seems like it might be useful to find out what is slow.  Is it the checking that the cache is up to date or the reading from the cache.  I have no idea how easy it would be to get the separate timing info.

Ben, you had an archiva instance in your room at one point.  I wonder how the timing compares to archiva on a local (not AWI) machine, for the same set of files.

**Comment by wandell** [comment]
From Stanford - similar to the numbers from NC, though a bit slower in both fetch and local

fetch times
    5.1849         0         0
    5.1849    1.5614         0
    5.1849    1.5614    1.5099

local load times
    0.0419         0         0
    0.0419    0.0428         0
    0.0419    0.0428    0.0417

**Comment by benjamin-heasly** [comment]
I put the Matlab profiler on it.  From the Matlab point of view, it's Gradle that's slow.  I think Gradle accounted for 99% of the time spent.

So to learn more about what's slow I think we need to profile the JVM instance running Gradle.  This should be doable.  I have used [VisualVM](https://visualvm.java.net/) to do JVM profiling before.

**Comment by benjamin-heasly** [comment]
I should mention -- the brainard-archiva instance and data are in Amazon's `us-west-2` region (aka "Oregon").  So I would expect the first-time fetches to be shortest from Stanford.

I think this was the case, since Brian reported 5.1849s whereas Nicolas and I reported 5.8078s and 35.9274s.

If this is not consistently the case, maybe other factors are dominating.  For example, my time of 35.9274s came from my home WiFi, which is not all that fancy.

**Comment by benjamin-heasly** [comment]
I did some testing with VisualVM and reading the gradle log (which includes timestamps for various internal things).

TLDR: I don't know a way to get the fetch times smaller than 1-2s.

Here is my understanding.

The Gradle daemon is a Java process that sits there waiting for us to request things.  I attached VisualVM to this process and also tailed its log file.  Then I ran our `cacheTests.m` script.  It looks like:
- the actual time spent performing each fetch was only 200-250ms
- most of the time was spent waiting between fetches

So why does it take us 1-2s per fetch?  Why can't we fetch every 250ms?  As far as I can tell, the extra time is overhead intrinsic to each Gradle request.

What happens when we invoke Gradle is that Gradle starts a short-lived Java process that submits our request to to the running Gradle daemon process.  This is good because some of the startup cost can be paid once and saved in the daemon's memory.  But it doesn't remove the startup cost for the requesting process itself.  This seems to be about 1-2s.

To test this understanding, I tried invoking Gradle with a trivial `--version` request.  I did this in bash:

```
$ cd ~/isetbio/RemoteDataToolbox/external/gradle
$ time ./gradlew --version

------------------------------------------------------------
Gradle 2.6
------------------------------------------------------------

Build time:   2015-08-10 13:15:06 UTC
Build number: none
Revision:     233bbf8e47c82f72cb898b3e0a96b85d0aad166e

Groovy:       2.3.10
Ant:          Apache Ant(TM) version 1.9.3 compiled on December 23 2013
JVM:          1.8.0_66 (Oracle Corporation 25.66-b17)
OS:           Linux 3.16.0-38-generic amd64


real    0m0.974s
user    0m1.643s
sys 0m0.075s
```

So we're not fetching anything, and still we have to wait 1643ms.  Alas.

**Comment by DavidBrainard** [comment]
The conclusion might be that RDTB is for big files where the overhead isn't such a large cost.  Indeed, the reason we started down this path is because we felt that we were filling up gitHub archives with some of the data files we were uploading.

It's possible that we want to keep small files with the code, and perhaps have RDTB know how to find such local files quickly -- if it's local (not in the cache, but on some search path) then snag it, otherwise get it using grade.  For example, we may want to put the FAST validation files back into the repository.  These are small as they just contain hash codes.

**Comment by benjamin-heasly** [comment]
I am closing this issue because it's old.  If we need to do more work in this area, let's create a new issue.
